pkg/cloudprovider/openstack/openstack.go

```diff
@@ -652,3 +655,153 @@ func (os *OpenStack) GetZone() (cloudprovider.Zone, error) {
+	disk, err := os.getVolume(partialDiskId)
+	if err != nil {
+		return err
+	}
+
+	cClient, err := openstack.NewComputeV2(os.provider, gophercloud.EndpointOpts{
+		Region: os.region,
+	})
+
+	compute_id, err := os.getComputeIDbyHostname(cClient)
+	if err != nil || compute_id == "" {
+		glog.Errorf("Unable to get minion's ip while detaching disk")
+		return err
+	}
+	if len(disk.Attachments) > 0 && disk.Attachments[0]["server_id"] != nil && compute_id == disk.Attachments[0]["server_id"] {
+		err = volumeattach.Delete(cClient, compute_id, disk.ID).ExtractErr()
+		if err != nil {
+			glog.Errorf("Failed to delete volume %s from compute %s attached %v\n", disk.ID, compute_id, err)
+			return err
+		}
+	} else {
+		errMsg := fmt.Sprintf("Disk: %s has no attachments or is not attached to compute: %s\n", disk.Name, compute_id)
+		glog.Errorf(errMsg)
+		return errors.New(errMsg)
+	}
+	return nil
+}
+
+//Takes a partial/full disk id or diskname
+
+func (os *OpenStack) getVolume(diskName string) (volumes.Volume, error) {
+	sClient, err := openstack.NewBlockStorageV1(os.provider, gophercloud.EndpointOpts{
+		Region: os.region,
+	})
+
+	var volume volumes.Volume
+	if err != nil || sClient == nil {
+		glog.Errorf("Unable to initialize cinder client for region: %s", os.region)
+		return volume, err
+	}
+
+	err = volumes.List(sClient, nil).EachPage(func(page pagination.Page) (bool, error) {
+		vols, err := volumes.ExtractVolumes(page)
+		if err != nil {
+			glog.Errorf("Failed to extract volumes: %v\n", err)
+			return false, err
+		} else {
+			for _, v := range vols {
+				glog.Infof("%s\t%s\t%v\n", v.ID, v.Name, v.Attachments)
+				if v.Name == diskName || strings.Contains(v.ID, diskName) {
+					volume = v
+					return true, nil
+				}
+			}
+		}
+		//if it reached here then no disk with the given name was found.
+		errmsg := fmt.Sprintf("Unable to find disk: %s in region %s", diskName, os.region)
+		return false, errors.New(errmsg)
+	})
+	if err != nil {
+		glog.Errorf("Error occured getting volume: %s\n", diskName)
+		return volume, err
+	}
+	return volume, err
+}
+
+func (os *OpenStack) getComputeIDbyHostname(cClient *gophercloud.ServiceClient) (string, error) {
+
+	hostname, err := ossys.Hostname()
+
+	if err != nil {
+		glog.Errorf("Failed to get Minion's hostname: %v\n", err)
+		return "", err
+	}
+
+	i, ok := os.Instances()
+	if !ok {
+		glog.Errorf("Unable to get instances")
+		return "", errors.New("Unable to get instances")
+	}
+
+	srvs, err := i.List(".")
+	if err != nil {
+		glog.Errorf("Failed to list servers: %v", err)
+		return "", err
+	}
+
+	if len(srvs) == 0 {
+		glog.Errorf("Found no servers in the region")
+		return "", errors.New("Found no servers in the region")
+	}
+	glog.Infof("found servers: %v\n", srvs)
+
+	for _, srvname := range srvs {
+		server, err := getServerByName(cClient, srvname)
+		if err != nil {
+			return "", err
+		} else {
+			if server.Metadata["hostname"] != nil && server.Metadata["hostname"] == hostname {
```

Comment:
@jsafrane thanks for testing it on Icehouse. I had originally developed it on havana. I will look for other ways to get the node id independent of metadata.